module/input/Microphone/src/Microphone.cpp

```diff
@@ -0,0 +1,359 @@
+        (char*) env_path, 
+        NULL 
+    };
+    
+    if((pipe(stdout_pipe) == -1) || (pipe(stderr_pipe) == -1) || (pipe(stdin_pipe) == -1)) {
+        goto cleanup;
+    }
+    
+    posix_spawn_file_actions_init(&actions);
+    
+    posix_spawn_file_actions_addclose(&actions, stdout_pipe[0]);
+    posix_spawn_file_actions_addclose(&actions, stderr_pipe[0]);
+    posix_spawn_file_actions_addclose(&actions, stdin_pipe[1]);
+
+    posix_spawn_file_actions_adddup2(&actions, stdout_pipe[1], STDOUT_FILENO);
+    posix_spawn_file_actions_adddup2(&actions, stderr_pipe[1], STDERR_FILENO);
+    posix_spawn_file_actions_adddup2(&actions, stdin_pipe[0], STDIN_FILENO);
+    
+    posix_spawn_file_actions_addclose(&actions, stdout_pipe[1]);
+    posix_spawn_file_actions_addclose(&actions, stderr_pipe[1]);
+    posix_spawn_file_actions_addclose(&actions, stdin_pipe[0]);
+        
+    if(posix_spawnp(&pid, args[0], &actions, 0, args, envp) != 0) {
+        goto cleanup;
+    }
+    
+    success = true;
+cleanup:
+    if(stdout_pipe[1]) close(stdout_pipe[1]);
+    if(stderr_pipe[1]) close(stderr_pipe[1]);
+    if(stdin_pipe[0]) close(stdin_pipe[0]);
+
+    *handles = {stdout_pipe[0], stderr_pipe[0], stdin_pipe[1]};
+    
+    if(actions.__actions) {
+        posix_spawn_file_actions_destroy(&actions);
+    }
+    if(env_path) {
+        free(env_path);
+    }
+    
+    return success;
+}
+
+//python -m voice2json --base-directory /home/nubots/voice2json/ -p en transcribe-wav
+bool voice2json_transcribe_stream(MicProcHandles *handles) {
+    char *const args[] = {
+        (char*) "python",     
+        (char*) "-m",
+        (char*) "voice2json",
+        (char*) "--base-directory",
+        (char*) "/home/nubots/voice2json/",
+        (char*) "-p",
+        (char*) "en",
+        (char*) "transcribe-stream",
+        NULL
+    };
+    
+    return spawn_process(handles, args);
+}
+
+//python -m voice2json --base-directory /home/nubots/voice2json/ -p en transcribe-wav --input-size
+bool voice2json_transcribe_wav(MicProcHandles *handles) {
+    char *const args[] = {
+        (char*) "python",     
+        (char*) "-m",
+        (char*) "voice2json",
+        (char*) "--base-directory",
+        (char*) "/home/nubots/voice2json/",
+        (char*) "-p",
+        (char*) "en",
+        (char*) "transcribe-wav",
+        (char*) "--input-size",
+        NULL
+    };
+    
+    return spawn_process(handles, args);
+}
+
+char *voice2json_recognize_intent(char *input) {
+    char *output = 0;
+    char buffer[0x1000];
+    ssize_t bytes_read = 0;
+    
+    char *const args[] = {
+        (char*) "python",     
+        (char*) "-m",
+        (char*) "voice2json",
+        (char*) "--base-directory",
+        (char*) "/home/nubots/voice2json/",
+        (char*) "-p",
+        (char*) "en",
+        (char*) "recognize-intent",
+        NULL
+    };
+    
+    MicProcHandles handles = {};
+    if(!spawn_process(&handles, args)) {
+        goto cleanup;
```

Comment:
did you write this? `goto` :eyes: :eyes:

Reply:
[git blame **/src/Microphone.cpp](https://xkcd.com/292/) 😉

Reply:
guilty as charged